### PR TITLE
Create dockerconfig in secret to login dockerhub registry

### PR DESCRIPTION
**Pull request by titanlien**
Verified with add, update and delete `kops_dockerconfig_creds`.
```
kops_dockerconfig_creds:
  user: 'user'
  token: 'token'
 ```
comming release v1.19.0

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -114,3 +114,13 @@ kops_default_kubeControllerManager:
     - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
 kops_default_kubeScheduler:
   enableProfiling: false
+
+###
+### Verify dockerconfig via get ratelimited
+### https://docs.docker.com/docker-hub/download-rate-limit/#how-can-i-check-my-current-rate
+###
+kops_default_dockerconfig_creds:
```
Comment (cytopia) [inline defaults/main.yml:122]: This should also be added to the Readme in: https://github.com/Flaconi/ansible-role-kops#additional-variables

**Q:** what happens when somebody does not want to specify a docker configuration and want to continue to use anonymous auth?

File: tasks/asserts.yml [hunk 2]
```diff
@@ -137,6 +137,22 @@
...
+  assert:
+    that:
+      - item.user is defined
+      - item.user | length > 0
+      - item.token is defined
+      - item.token | length > 0
+    msg: "kops_dockerconfig_creds should include user and token and validated in dockerhub"
+  with_items:
+    - >-
+      {%- if kops_dockerconfig_creds is defined -%}
```
Comment (cytopia) [inline tasks/asserts.yml:150]: Where is `kops_dockerconfig_creds` defined? And where can I find documentation about this?

**Review by cytopia — CHANGES_REQUESTED** [review]

**Comment by cytopia** [comment]
Please also add all available configuration options to this section: https://github.com/Flaconi/ansible-role-kops#fully-customized

File: defaults/main.yml (hunk 1 above)
Comment (titanlien) [reply defaults/main.yml:122]: > Question:

They just don't defined `kops_dockerconfig_creds` this map, then they will go with anonymous user.
`kops_default_dockerconfig_creds`: I will update it in readme

File: tasks/generate_templates.yml [hunk 3]
```diff
@@ -16,6 +16,40 @@
     - encryption-config.yml
     - ssh-key.pub
 
+- name: "({{ cluster.name }}) verify kops_dockerconfig_creds for dockerhub"
+  uri:
+    url: "{{ kops_dockerconfig_creds.url | default(kops_default_dockerconfig.url )}}"
```
Comment (cytopia) [inline tasks/generate_templates.yml:21]: I don't think we need a default here as this piece of code is never reached. If `kops_dockerconfig_creds` is not defined, this block will not be executed.

File: defaults/main.yml [hunk 4]
```diff
@@ -114,3 +114,13 @@ kops_default_kubeControllerManager:
     - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
 kops_default_kubeScheduler:
   enableProfiling: false
+
+###
+### Verify dockerconfig via get ratelimited
+### https://docs.docker.com/docker-hub/download-rate-limit/#how-can-i-check-my-current-rate
+###
+kops_default_dockerconfig:
```
Comment (cytopia) [inline defaults/main.yml:122]: This defines a default, but when is this default ever used?

File: tasks/asserts.yml (hunk 2 above)
Comment (titanlien) [reply tasks/asserts.yml:150]: Updated it in README

File: tasks/asserts.yml [hunk 5]
```diff
@@ -137,6 +137,22 @@
...
+      {{ kops_dockerconfig_creds }}
+      {%- else -%}
+      {{ kops_default_dockerconfig }}
```
Comment (cytopia) [inline tasks/asserts.yml:153]: Not sure if the else block is needed at all.

File: tasks/generate_templates.yml (hunk 3 above)
Comment (titanlien) [reply tasks/generate_templates.yml:21]: We still need `url` in default until we want to access private registry. Then we don't define `url` in `kops_dockerconfig_creds` in vault when we try to login dockerhub.

File: tasks/asserts.yml (hunk 5 above)
Comment (titanlien) [reply tasks/asserts.yml:153]: Update it with new format, `loop`. https://docs.ansible.com/ansible/devel/porting_guides/porting_guide_2.5.html#migrating-from-with-x-to-loop

File: defaults/main.yml (hunk 4 above)
Comment (titanlien) [reply defaults/main.yml:122]: we still need the url and timeout value as default value

**Review by cytopia — APPROVED** [review]

**Pull request merged**